File: ai-service/app/schemas.py
"""Pydantic request/response models matching docs/14-api-ai-contracts.md."""

from __future__ import annotations

from typing import Literal

from pydantic import BaseModel, ConfigDict, Field


class HealthResponse(BaseModel):
    service: str
    status: str


class ErrorBody(BaseModel):
    code: str
    message: str


class ErrorResponse(BaseModel):
    error: ErrorBody


class AssistantActionProposal(BaseModel):
    model_config = ConfigDict(extra="forbid")

    id: str = Field(..., min_length=12, max_length=120)
    action: str = Field(..., min_length=1, max_length=80)
    title: str = Field(..., min_length=1, max_length=200)
    description: str = Field(..., min_length=1, max_length=1000)
    requiredPermission: str = Field(..., min_length=1, max_length=80)
    payload: dict[str, object] = Field(default_factory=dict)
    effect: str = Field(..., min_length=1, max_length=1000)


class AssistantContext(BaseModel):
    model_config = ConfigDict(extra="allow")

    pathname: str | None = Field(default=None, max_length=300)
    search: str | None = Field(default=None, max_length=800)
    filters: dict[str, object] = Field(default_factory=dict)


class AssistantToolDescriptor(BaseModel):
    model_config = ConfigDict(extra="forbid")

    name: str
    description: str
    requiredPermission: str | None = None
    mutates: bool
    inputSchema: dict[str, object] = Field(default_factory=dict)


class AssistantChatRequest(BaseModel):
    model_config = ConfigDict(extra="forbid")

    message: str = Field(..., min_length=1, max_length=2000)
    context: AssistantContext | None = None
    history: list[dict[str, str]] = Field(default_factory=list, max_length=10)
    permissions: list[str] = Field(default_factory=list)
    toolSessionToken: str = Field(..., min_length=24, max_length=160)
    toolsUrl: str = Field(..., min_length=1)


class AssistantChatResponse(BaseModel):
    model_config = ConfigDict(extra="forbid")

    type: Literal["answer", "navigation", "permission_denied", "action_proposal"]
    message: str = Field(..., min_length=1, max_length=4000)
    path: str | None = None
    requiredPermission: str | None = None
    proposal: AssistantActionProposal | None = None


class PublicChatTurn(BaseModel):
    model_config = ConfigDict(extra="forbid")

    role: Literal["user", "assistant"]
    content: str = Field(..., min_length=1, max_length=4500)


class PublicChatRequest(BaseModel):
    model_config = ConfigDict(extra="forbid")

    message: str = Field(..., min_length=1, max_length=500)
    language: Literal["en", "sw"] | None = None
    turns: list[PublicChatTurn] = Field(default_factory=list, max_length=6)
    toolSessionToken: str = Field(..., min_length=24, max_length=160)
    toolsUrl: str = Field(..., min_length=1)


class PublicChatResponse(BaseModel):
    model_config = ConfigDict(extra="forbid")

    answer: str = Field(..., min_length=1, max_length=4000)

File: api/src/modules/assistant/schemas.ts
/**
 * App assistant request contracts.
 * Natural-language parsing is intentionally bounded; execution remains API-owned.
 */

import { z } from 'zod'

const contextSchema = z
  .object({
    pathname: z.string().trim().max(300).optional(),
    search: z.string().trim().max(800).optional(),
    filters: z.record(z.string(), z.unknown()).optional(),
  })
  .passthrough()
  .optional()

export const assistantMessageBodySchema = z
  .object({
    message: z
      .string()
      .trim()
      .min(1, 'message is required')
      .max(2000, 'message must be at most 2000 characters'),
    context: contextSchema,
    history: z.array(z.object({
      role: z.enum(['user', 'assistant']),
      content: z.string().trim().min(1).max(2000),
    })).max(10).optional().default([]),
  })
  .strict()

export type AssistantMessageBody = z.infer<typeof assistantMessageBodySchema>

export const assistantActionParamSchema = z.object({
  id: z.string().trim().min(12, 'Action id is invalid').max(120),
})

export type AssistantActionParam = z.infer<typeof assistantActionParamSchema>

const permissionCodeSchema = z.string().trim().min(1).max(80)

export const assistantActionProposalSchema = z.object({
  id: z.string().trim().min(12).max(120),
  action: z.string().trim().min(1).max(80),
  title: z.string().trim().min(1).max(200),
  description: z.string().trim().min(1).max(1000),
  requiredPermission: permissionCodeSchema,
  payload: z.record(z.string(), z.unknown()),
  effect: z.string().trim().min(1).max(1000),
})

export const assistantResultSchema = z.discriminatedUnion('type', [
  z.object({
    type: z.literal('answer'),
    message: z.string().trim().min(1).max(4000),
  }),
  z.object({
    type: z.literal('navigation'),
    message: z.string().trim().min(1).max(1000),
    path: z.string().trim().min(1).max(300),
    requiredPermission: permissionCodeSchema.optional(),
  }),
  z.object({
    type: z.literal('permission_denied'),
    message: z.string().trim().min(1).max(1000),
    requiredPermission: permissionCodeSchema,
  }),
  z.object({
    type: z.literal('action_proposal'),
    message: z.string().trim().min(1).max(1000),
    proposal: assistantActionProposalSchema,
  }),
])

export type AssistantResultBody = z.infer<typeof assistantResultSchema>

export const assistantChatRequestSchema = z.object({
  message: z.string().trim().min(1).max(2000),
  context: contextSchema,
  history: z.array(z.object({ role: z.enum(['user', 'assistant']), content: z.string().trim().min(1).max(2000) })).max(10).optional().default([]),
  permissions: z.array(permissionCodeSchema).max(100),
  toolSessionToken: z.string().trim().min(24).max(160),
  toolsUrl: z.string().url(),
})

export type AssistantChatRequest = z.infer<typeof assistantChatRequestSchema>

export const assistantToolListResponseSchema = z.object({
  tools: z.array(
    z.object({
      name: z.string().trim().min(1).max(120),
      description: z.string().trim().min(1).max(1000),
      requiredPermission: permissionCodeSchema.optional(),
      mutates: z.boolean(),
      inputSchema: z.record(z.string(), z.unknown()),
    }),
  ),
})

export const assistantToolCallBodySchema = z.object({
  name: z.string().trim().min(1).max(120),
  arguments: z.record(z.string(), z.unknown()).optional().default({}),
})

export type AssistantToolCallBody = z.infer<typeof assistantToolCallBodySchema>
